lib/filtrex/condition.ex

```diff
@@ -137,6 +159,10 @@ defmodule Filtrex.Condition do
     Application.get_env(:filtrex, :conditions, @modules)
   end
 
+  defp put_dump_value(condition) do
+    Map.update!(condition, :value, &(condition.__struct__.dump_value(&1)))
```

Comment:
I like the consistency of `dump_value` but I think we should extract it into a protocol (e.g. `Filtrex.Encoder.Map`) since it'll need to be implemented if a new condition is desired. We'll also want to insert the `@behaviour` into `use` macro for `Filtrex.Condition` such that it will warn the developer who adds a condition without specifying an `dump_map_value` function.

Reply:
Using behaviors with protocols looks strange to me. I also wasn't able to find any examples of this.

Reply:
This library actually has an example in `Filtrex.Condition` with the `@callback` indicators and the `@behaviour` that is inserted into a given module when the `use` macro is incorporated.

Reply:
Sorry for the long delay! I'm probably missing something but aren't we talking about protocols? As I see in `Filtrex.Condition`, all `@callback` has no relation to any existing protocols?

Reply:
Yes. Protocols.

Reply:
Yeah, so I moved this logic into a protocol. As I can see there is no way to force implementation of protocol with behavior, but we can implement a default implementation which will raise a warning or error.

Reply:
I was not able to achieve the desired behavior with the default implementation, so I had to fall back to the after-compile hook

Reply:
Excellent. 👌